### DISCUSSION TITLE
Rivet as a Fixture in E2E Tests

### DISCUSSION THREAD

**Discussion by alleneubank** [paradigmxyz/rivet#36] [Ideas]
# Objective

To augment Rivet's functionality by allowing it to act as a test fixture in E2E tests, specifically targeting frameworks like Cypress and Playwright.

## Proposal

What if we could have Rivet's provider injected directly into the test page? This would allow for easy setup and teardown without the need for Rivet to run as a browser extension. The injected provider would be capable of communicating with Rivet over HTTP or through page handles provided by the testing frameworks.

By enabling provider injection, we could set up and manipulate the blockchain state as part of our test suite. This opens the door for testing more complex interactions and workflows in dApps and smart contracts.

The aim is twofold—integrating seamlessly with existing testing frameworks and offering the ability to validate complex blockchain use-cases. The beauty lies in its simplicity and the wide range of testing scenarios it would unlock.

**Comment by alleneubank** [comment]
For those coming here, this library works great so far, https://github.com/cawabunga/headless-web3-provider

Possibly not needed with rivet since most of the work is already done by the provider above.

**Reply by jxom** [reply]
Looks good! But I think there is a need for a headless injected provider with Anvil + Viem! Would be cool for someone to look into running Rivet headlessly.

**Reply by alleneubank** [reply]
Definitely! When I looked at this last, I was going to propose a new entry point and package export named fixture. A good starting point would be to a merge version of the background entry point and the inpage entrypoint that exposes methods for approving requests.

Or it could just follow `headless-web3-provider` and use viem instead of ethers. There's probably a lot of code in rivet that could be re-used for the injected provider.
